http.go

```diff
@@ -0,0 +1,172 @@
+package stream_chat
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+type Error struct {
+	Code            int               `json:"code"`
+	Message         string            `json:"message"`
+	ExceptionFields map[string]string `json:"exception_fields,omitempty"`
+	StatusCode      int               `json:"StatusCode"`
+	Duration        string            `json:"duration"`
+	MoreInfo        string            `json:"more_info"`
+
+	RateLimit *RateLimit `json:"-"`
+}
+
+func (e Error) Error() string {
+	return e.Message
+}
+
+const (
+	HeaderRateLimit     = "X-Ratelimit-Limit"
+	HeaderRateRemaining = "X-Ratelimit-Remaining"
+	HeaderRateReset     = "X-Ratelimit-Reset"
+)
+
+type RateLimit struct {
+	Reset     time.Time
+	Limit     int
+	Remaining int
+}
+
+func NewRateLimitFromHeaders(headers http.Header) *RateLimit {
+	var rl RateLimit
+
+	limit, err := strconv.Atoi(headers.Get(HeaderRateLimit))
+	if err == nil {
+		rl.Limit = limit
+	}
+	remaining, err := strconv.Atoi(headers.Get(HeaderRateRemaining))
+	if err == nil {
+		rl.Remaining = remaining
+	}
+	reset, err := strconv.ParseInt(headers.Get(HeaderRateReset), 10, 64)
+	if err == nil && reset > 0 {
+		rl.Reset = time.Unix(reset, 0)
+	}
+
+	return &rl
+}
+
+func (c *Client) parseResponse(resp *http.Response, result interface{}) error {
+	if resp.Body == nil {
+		return errors.New("http body is nil")
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		panic(err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 399 {
+		var apiErr Error
+		err := json.Unmarshal(b, &apiErr)
+		if err != nil {
+			// IP rate limit errors sent by our Edge infrastructure are not JSON encoded.
+			// If decode fails here, we need to handle this manually.
+			apiErr.Message = string(b)
+			apiErr.StatusCode = resp.StatusCode
+			return apiErr
+		}
+
+		if resp.StatusCode == http.StatusTooManyRequests {
```

Comment:
we have this information in every response so every error, it might help somewhere